### DISCUSSION TITLE
why is there a varlogs target?

### DISCUSSION THREAD

**Discussion by tcurdt** [grafana-cold-storage/agent#2672] [Q&A]
Why does a log scraper have a target "localhost"?

If it is for labeling, "localhost" wouldn't be useful.
And otherwise there is no host or port involved.

        - job_name: varlogs
          static_configs:
            - targets: [localhost]
              labels:
                job: varlogs
                __path__: /var/log/*log

I am a little confused.
Could you please clarify?

**Comment by rfratto** [comment]
It's an artifact of using Prometheus Service Discovery (which is network-based) for log collection. The `localhost` address gets ignored, but there must be some address to be able to generate targets to collect logs from.

**Reply by tcurdt** [reply]
So it is not really ignored but used as sort of a reference?

**Reply by rfratto** [reply]
I'm not sure what you mean by "reference" here, but it's just necessary to emit targets at all using `static_configs`. Log collection needs targets to work, and it relies on Prometheus Service Discovery to emit those targets. 

`static_configs` will emit one target per entry in the `targets` array, so it must be populated with _something_ so that the log collection has a target to work with. As far as I can see in the original code from Promtail, the address is never read.

**Reply by tcurdt** [reply]
I read "must be some address to be able to generate targets to collect logs from" as there needs to be something to reference the data by. But if it's not even read that can't really be it.
Just feels a bit weird.

I guess you try to stay close to the Promtail/Prometheus roots instead of coming with a custom syntax?
Anyway - thanks for the clarification.

**Reply by rfratto** [reply]
> I guess you try to stay close to the Promtail/Prometheus roots instead of coming with a custom syntax?

The logging code in the static agent mode (the default) is literally Promtail, so yes :) Notably an address no longer required for log collection in [Flow](https://grafana.com/blog/2022/09/29/introducing-programmable-pipelines-with-grafana-agent-flow/?pg=blog&plcmt=body-txt) mode [since v0.30](https://grafana.com/blog/2022/12/20/grafana-agent-v0.30-flow-adds-support-for-logging-pipelines-and-graduates-to-beta/): 

```river 
discovery.file "varlogs" {
  path_targets = [{
    job      = "varlogs",
    __path__ = "/var/log/*log"
  }]
}

loki.source.file "default" {
    targets    = discovery.file.varlogs.targets
    forward_to = [...]
}
```
